mountpoint-s3-crt/src/s3/endpoint_resolver.rs

```diff
@@ -0,0 +1,251 @@
+        //`aws_endpoint_resolved_enpoint_get_url` ensures to return an initialised aws_byte_cursor for url or else it will return an error.
+        unsafe {
+            aws_endpoints_resolved_endpoint_get_url(self.inner.as_ptr(), url.as_mut_ptr()).ok_or_last_error()?;
+            let uri = aws_byte_cursor_as_slice(&url.assume_init());
+            let uri = OsStr::from_bytes(uri);
+            Uri::new_from_str(allocator, uri)
+        }
+    }
+}
+
+impl Drop for ResolvedEndpoint {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `aws_endpoints_resolved_endpoint`, and on Drop it's safe to decrement
+        // the reference count since we won't use it again through `self.`
+        unsafe {
+            aws_endpoints_resolved_endpoint_release(self.inner.as_ptr());
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::ffi::OsStr;
+
+    use crate::common::allocator::Allocator;
+
+    use super::{RequestContext, RuleEngine};
+
+    #[test]
+    fn test_regions_outside_aws_partition() {
+        let mut new_allocator = Allocator::default();
+        let endpoint_rule_engine = RuleEngine::new(&new_allocator);
+        let mut endpoint_request_context = RequestContext::new(&new_allocator);
+        endpoint_request_context
+            .add_string(&new_allocator, OsStr::new("Bucket"), OsStr::new("s3-bucket-test"))
+            .expect("Should set bucket name");
+        endpoint_request_context
+            .add_string(&new_allocator, OsStr::new("Region"), OsStr::new("cn-north-1"))
+            .expect("Should set region name");
+        let endpoint_resolved = endpoint_rule_engine
+            .resolve(endpoint_request_context)
+            .expect("Endpoint should be resolved");
+        let endpoint_uri = endpoint_resolved
+            .get_url(&mut new_allocator)
+            .expect("Unable to get the Uri from resolved endpoint");
+        assert_eq!(
+            endpoint_uri.as_os_str(),
+            "https://s3-bucket-test.s3.cn-north-1.amazonaws.com.cn"
+        );
+    }
+
+    #[test]
+    fn test_fips_setting() {
+        let mut new_allocator = Allocator::default();
+        let endpoint_rule_engine = RuleEngine::new(&new_allocator);
+        let mut endpoint_request_context = RequestContext::new(&new_allocator);
+        endpoint_request_context
+            .add_string(&new_allocator, OsStr::new("Bucket"), OsStr::new("s3-bucket-test"))
+            .expect("Should set bucket name");
+        endpoint_request_context
+            .add_string(&new_allocator, OsStr::new("Region"), OsStr::new("us-east-1"))
+            .expect("Should set region name");
+        endpoint_request_context
+            .add_boolean(&new_allocator, OsStr::new("UseFIPS"), true)
+            .expect("Should set to use FIPS");
+        let endpoint_resolved = endpoint_rule_engine
+            .resolve(endpoint_request_context)
+            .expect("Endpoint should be resolved");
```

Comment:
For most of these, I think `.unwrap()` would be fine. The expect message doesn't tell us anything for most.

For the last one, why do we expect the endpoint resolution to work?

```suggestion
        endpoint_request_context
            .add_string(&new_allocator, OsStr::new("Bucket"), OsStr::new("s3-bucket-test"))
            .unwrap();
        endpoint_request_context
            .add_string(&new_allocator, OsStr::new("Region"), OsStr::new("us-east-1"))
            .unwrap();
        endpoint_request_context
            .add_boolean(&new_allocator, OsStr::new("UseFIPS"), true)
            .unwrap();
        let endpoint_resolved = endpoint_rule_engine
            .resolve(endpoint_request_context)
            .expect("endpoint should resolve as rules should match context");
```